File: app/src/main/java/com/quran/labs/androidquran/presenter/quran/ayahtracker/AyahImageTrackerItem.java
package com.quran.labs.androidquran.presenter.quran.ayahtracker;

import android.content.Context;
import android.graphics.RectF;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.quran.data.core.QuranInfo;
import com.quran.data.model.SuraAyah;
import com.quran.data.model.selection.SelectedAyahPosition;
import com.quran.labs.androidquran.data.QuranDisplayData;
import com.quran.labs.androidquran.ui.helpers.HighlightType;
import com.quran.labs.androidquran.ui.helpers.QuranDisplayHelper;
import com.quran.labs.androidquran.ui.util.ImageAyahUtils;
import com.quran.labs.androidquran.util.QuranUtils;
import com.quran.labs.androidquran.view.HighlightingImageView;
import com.quran.page.common.data.AyahBounds;
import com.quran.page.common.data.AyahCoordinates;
import com.quran.page.common.data.PageCoordinates;
import com.quran.page.common.draw.ImageDrawHelper;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class AyahImageTrackerItem extends AyahTrackerItem {
  private final QuranInfo quranInfo;
  private final QuranDisplayData quranDisplayData;
  private final boolean isPageOnRightSide;
  private final int screenHeight;
  private final Set<ImageDrawHelper> imageDrawHelpers;
  @Nullable Map<String, List<AyahBounds>> coordinates;
  @NonNull final HighlightingImageView ayahView;


  public AyahImageTrackerItem(int page,
                              int screenHeight,
                              QuranInfo quranInfo,
                              QuranDisplayData quranDisplayData,
                              @NonNull Set<ImageDrawHelper> imageDrawHelpers,
                              @NonNull HighlightingImageView highlightingImageView) {
    this(page, screenHeight, quranInfo, quranDisplayData, false, imageDrawHelpers, highlightingImageView);
  }

  public AyahImageTrackerItem(int page,
                              int screenHeight,
                              QuranInfo quranInfo,
                              QuranDisplayData quranDisplayData,
                              boolean isPageOnTheRight,
                              @NonNull Set<ImageDrawHelper> imageDrawHelpers,
                              @NonNull HighlightingImageView highlightingImageView) {
    super(page);
    this.ayahView = highlightingImageView;
    this.quranInfo = quranInfo;
    this.quranDisplayData = quranDisplayData;
    this.screenHeight = screenHeight;
    this.imageDrawHelpers = imageDrawHelpers;
    this.isPageOnRightSide = isPageOnTheRight;
  }

  @Override
  public void onSetPageBounds(PageCoordinates pageCoordinates) {
    final int page = getPage();
    if (page == pageCoordinates.getPage()) {
      // this is only called if overlayText is set
      final RectF pageBounds = pageCoordinates.getPageBounds();
      if (!pageBounds.isEmpty()) {
        ayahView.setPageBounds(pageBounds);
        Context context = ayahView.getContext();
        String suraText = quranDisplayData.getSuraNameFromPage(context, page, true);
        String juzText = quranDisplayData.getJuzDisplayStringForPage(context, page);
        String pageText = QuranUtils.getLocalizedNumber(context, page);
        String rub3Text = QuranDisplayHelper.displayRub3(context, quranInfo, page);
        ayahView.setOverlayText(context, suraText, juzText, pageText, rub3Text);
      }
      ayahView.setPageData(pageCoordinates, imageDrawHelpers);
    }
  }

  @Override
  public void onSetAyahCoordinates(AyahCoordinates ayahCoordinates) {
    final int page = getPage();
    if (page == ayahCoordinates.getPage()) {
      this.coordinates = ayahCoordinates.getAyahCoordinates();
      if (!coordinates.isEmpty()) {
        ayahView.setAyahData(ayahCoordinates);
        ayahView.invalidate();
      }
    }
  }

  @Override
  public boolean onHighlightAyah(int page, int sura, int ayah, @NonNull HighlightType type, boolean scrollToAyah) {
    if (getPage() == page && coordinates != null) {
      ayahView.highlightAyah(sura, ayah, type);
      ayahView.invalidate();
      return true;
    } else if (coordinates != null) {
      ayahView.unHighlight(type);
    }
    return false;
  }

  @Override
  public void onHighlightAyat(int page, @NonNull Set<String> ayahKeys, @NonNull HighlightType type) {
    if (getPage() == page) {
      ayahView.highlightAyat(ayahKeys, type);
      ayahView.invalidate();
    }
  }

  @Override
  public void onUnHighlightAyah(int page, int sura, int ayah, @NonNull HighlightType type) {
    if (getPage() == page) {
      ayahView.unHighlight(sura, ayah, type);
    }
  }

  @Override
  public void onUnHighlightAyahType(@NonNull HighlightType type) {
    ayahView.unHighlight(type);
  }

  @Override
  public SelectedAyahPosition getToolBarPosition(int page, int sura, int ayah, int toolBarWidth,
                                                     int toolBarHeight) {
    if (getPage() == page) {
      final List<AyahBounds> bounds = coordinates == null ? null :
          coordinates.get(sura + ":" + ayah);
      final int screenWidth = ayahView.getWidth();
      if (bounds != null && screenWidth > 0) {
        SelectedAyahPosition position =
            ImageAyahUtils.getToolBarPosition(bounds, ayahView.getImageMatrix(),
                screenWidth, screenHeight, toolBarWidth, toolBarHeight);

        final int topPadding = ayahView.getPaddingTop();
        if (topPadding > 0) {
          position = position.withY(position.getY() + topPadding);
        }

        if (isPageOnRightSide) {
          // need to adjust offset because our x is really x plus one page
          return position.withX(position.getX() + ayahView.getWidth());
        }
        return position;
      }
    }
    return null;
  }

  @Nullable
  @Override
  public SuraAyah getAyahForPosition(int page, float x, float y) {
    return getPage() == page ?
        ImageAyahUtils.getAyahFromCoordinates(coordinates, ayahView, x, y) : null;
  }
}

File: app/src/main/java/com/quran/labs/androidquran/presenter/quran/ayahtracker/AyahImageTrackerItem.kt
package com.quran.labs.androidquran.presenter.quran.ayahtracker

import android.view.View
import com.quran.data.core.QuranInfo
import com.quran.data.model.AyahGlyph
import com.quran.data.model.AyahWord
import com.quran.data.model.SuraAyah
import com.quran.data.model.highlight.HighlightType
import com.quran.data.model.selection.SelectionIndicator
import com.quran.labs.androidquran.data.QuranDisplayData
import com.quran.labs.androidquran.ui.helpers.QuranDisplayHelper
import com.quran.labs.androidquran.ui.util.ImageAyahUtils
import com.quran.labs.androidquran.util.QuranUtils
import com.quran.labs.androidquran.view.HighlightingImageView
import com.quran.page.common.data.AyahBounds
import com.quran.page.common.data.AyahCoordinates
import com.quran.page.common.data.PageCoordinates
import com.quran.page.common.data.coordinates.PageGlyphsCoords
import com.quran.page.common.draw.ImageDrawHelper

open class AyahImageTrackerItem @JvmOverloads constructor(
  page: Int,
  private val quranInfo: QuranInfo?,
  private val quranDisplayData: QuranDisplayData,
  private val isPageOnRightSide: Boolean = false,
  private val imageDrawHelpers: Set<ImageDrawHelper>,
  val ayahView: HighlightingImageView
) : AyahTrackerItem(page) {
  @JvmField var coordinates: Map<String, List<AyahBounds>>? = null
  @JvmField var pageGlyphsCoords: PageGlyphsCoords? = null

  override fun onSetPageBounds(pageCoordinates: PageCoordinates) {
    if (this.page == pageCoordinates.page) {
      // this is only called if overlayText is set
      val pageBounds = pageCoordinates.pageBounds
      if (!pageBounds.isEmpty) {
        ayahView.setPageBounds(pageBounds)
        val context = ayahView.context
        val suraText = quranDisplayData.getSuraNameFromPage(context, page, true)
        val juzText = quranDisplayData.getJuzDisplayStringForPage(context, page)
        val pageText = QuranUtils.getLocalizedNumber(context, page)
        val rub3Text = QuranDisplayHelper.displayRub3(context, quranInfo, page)
        val manzilText = quranDisplayData.getManzilForPage(context, page)
        ayahView.setOverlayText(suraText, juzText, pageText, rub3Text, manzilText)
      }
      ayahView.setPageData(pageCoordinates, imageDrawHelpers)
    }
  }

  override fun onSetAyahCoordinates(ayahCoordinates: AyahCoordinates) {
    if (this.page == ayahCoordinates.page) {
      val coordinates = ayahCoordinates.ayahCoordinates
      this.coordinates = coordinates
      this.pageGlyphsCoords = ayahCoordinates.glyphCoordinates
      if (coordinates.isNotEmpty()) {
        ayahView.setAyahData(ayahCoordinates)
        ayahView.invalidate()
      }
    }
  }

  override fun onHighlightAyah(
    page: Int,
    sura: Int,
    ayah: Int,
    word: Int,
    type: HighlightType,
    scrollToAyah: Boolean
  ): Boolean {
    if (this.page == page && coordinates != null) {
      ayahView.highlightAyah(sura, ayah, word, type)
      ayahView.invalidate()
      return true
    } else if (coordinates != null) {
      ayahView.unHighlight(type)
    }
    return false
  }

  override fun onHighlightAyat(page: Int, ayahKeys: Set<String>, type: HighlightType) {
    if (this.page == page) {
      ayahView.highlightAyat(ayahKeys, type)
      ayahView.invalidate()
    }
  }

  override fun onUnHighlightAyah(page: Int, sura: Int, ayah: Int, type: HighlightType) {
    if (this.page == page) {
      ayahView.unHighlight(sura, ayah, type)
    }
  }

  override fun onUnHighlightAyah(page: Int, sura: Int, ayah: Int, word: Int, type: HighlightType) {
    if (this.page == page) {
      ayahView.unHighlight(sura, ayah, word, type)
    }
  }

  override fun onUnHighlightAyahType(type: HighlightType) {
    ayahView.unHighlight(type)
  }

  override fun getToolBarPosition(page: Int, sura: Int, ayah: Int): SelectionIndicator {
    if (this.page == page) {
      val coordinates = coordinates
      val bounds = if (coordinates == null) null else coordinates["$sura:$ayah"]
      return getToolBarPosition(page, bounds)
    }
    return SelectionIndicator.None
  }

  override fun getToolBarPosition(page: Int, word: AyahWord): SelectionIndicator {
    if (this.page == page) {
      val bounds = pageGlyphsCoords?.getBoundsForWord(word)?.map { AyahBounds(-1, -1, it) }
      return getToolBarPosition(page, bounds)
    }
    return SelectionIndicator.None
  }

  override fun getToolBarPosition(page: Int, glyph: AyahGlyph): SelectionIndicator {
    if (this.page == page) {
      val bounds = pageGlyphsCoords?.getBoundsForGlyph(glyph)?.map { AyahBounds(-1, -1, it) }
      return getToolBarPosition(page, bounds)
    }
    return SelectionIndicator.None
  }

  private fun getToolBarPosition(page: Int, bounds: List<AyahBounds>?): SelectionIndicator {
    if (this.page == page) {
      val screenWidth = ayahView.width
      if (bounds != null && screenWidth > 0) {
        val yPadding = ayahView.paddingTop
        val xPadding = if (isPageOnRightSide) ayahView.width else 0
        return ImageAyahUtils.getToolBarPosition(bounds, ayahView.imageMatrix, xPadding, yPadding)
      }
    }
    return SelectionIndicator.None
  }

  override fun getAyahForPosition(page: Int, x: Float, y: Float): SuraAyah? {
    return if (this.page == page) ImageAyahUtils.getAyahFromCoordinates(
      coordinates,
      ayahView,
      x,
      y
    ) else null
  }

  override fun getGlyphForPosition(page: Int, x: Float, y: Float): AyahGlyph? {
    return if (this.page == page) {
      // Convert screen x,y coordinates -> page x,y coordinates
      val (pgX, pgY) = ImageAyahUtils.getPageXY(x, y, ayahView) ?: return null
      // Return the glyph at that page x,y coordinates
      pageGlyphsCoords?.getGlyphAtPoint(pgX, pgY)
    } else null
  }

  override fun getAyahView(): View {
    return ayahView
  }
}
